### ISSUE TITLE
[pubsub] MessageIterator's Stop with inflight messages causes KeepAliver's Stop to hang.

### ISSUE THREAD

**Issue by epsniff** [GoogleCloudPlatform/google-cloud-go#491]
Having any inflight messages (those NOT acked or nacked) result in unexpected behaviour with the pubsub iterator.   One of two things happen with inflight messages after one calls `MessageIterator.Stop()`:

- **a)** If one stops the readers by calling `MessageIterator.Stop()` without canceling the provided context.  Then the `KeepAliver` hangs waiting on it's waitgroup in `KeepAliver.Stop()`.   Which will not happen until all messages are Acked/Nacked and removed from the keepaliver.

- **b)** If one stops the readers by canceling the provided context and then calling `MessageIterator.Stop()`, then all ack's waiting in the `acker` queue are dropped and replay'ed after they expire.

Case b makes sense and is what I would expect.  Case a) is unexpected, as I would expect it to flush the acker (which it does) and then just kill the KeepAliver as it's no longer needed.

Thoughts?

I coded up a test case that reproduced the issue and my idea for fixing case a).   The code changes are just there to present the idea and are rough draft.   The basic idea is to use child context objects for the subcomponents so they can be shutdown in a controlled order.

https://github.com/epsniff/google-cloud-go/pull/2

**Comment by jba** [comment]
We'll be removing iterators soon, and the implementation won't match what you describe. However, behavior (a), which is the intended behavior, will remain. (b) will no longer be possible (though we can add it back if desired).

**Comment by epsniff** [comment]
The new design of the pub-sub SDK correctly documents that this is behavior, so I'm closing this ticket.
